File: inventory_stack/lib/core/models/migrations.dart
import 'package:inventory_stack/core/models/item.dart';
import 'package:inventory_stack/core/models/place.dart';
import 'package:json_annotation/json_annotation.dart';

part 'migrations.g.dart';

@JsonSerializable()
class MigrationsData {
  @JsonKey(name: "create_at")
  final DateTime createAt;
  @JsonKey(name: "upgrade_at")
  final DateTime upgradeAt;
  @JsonKey(name: "delete_at")
  final DateTime? deleteAt;
  final String uuid;
  @JsonKey(name: "item_uuid")
  final String itemUuid;
  @JsonKey(name: "from_uuid")
  final String fromUuid;
  @JsonKey(name: "to_uuid")
  final String toUuid;
  final ItemData? item;
  final PlaceData? from;
  final PlaceData? to;


  factory MigrationsData.fromJson(Map<String, dynamic> json) => _$MigrationsDataFromJson(json);

  MigrationsData(this.createAt, this.upgradeAt, this.deleteAt, this.uuid, this.itemUuid, this.fromUuid, this.toUuid, this.item, this.from, this.to);

  Map<String, dynamic> toJson() => _$MigrationsDataToJson(this);
}

File: server/db/models/migrations.go
package models

import (
	"time"

	"github.com/go-pg/pg/v10"
)

type Migration struct {
	UUID         string    `pg:"uuid,unique,pk" json:"uuid"`
	CreateAt     time.Time `pg:"create_at" json:"create_at"`
	UpgradeAt    time.Time `pg:"upgrade_at" json:"upgrade_at"`
	DeleteAt     time.Time `pg:"delete_at" json:"delete_at"`
	ReturnedDate time.Time `pg:"returned_date" json:"returned_date"`
	ItemUUID     string    `pg:"item_uuid" json:"item_uuid"`
	FromUUID     string    `pg:"from_uuid" json:"from_uuid"`
	ToUUID       string    `pg:"to_uuid" json:"to_uuid"`
	Item         *Item     `pg:"rel:has-one" json:"item"`
	From         *Place    `pg:"rel:has-one" json:"from"`
	To           *Place    `pg:"rel:has-one" json:"to"`
}

func (itm *Migration) CreateMigration(conn *pg.DB) error {
	if _, err := conn.Model(itm).Insert(itm); err != nil {
		return err
	}
	return nil
}

func (itm *Migration) GetAllMigrations(conn *pg.DB) (*[]Migration, error) {
	migrations := &[]Migration{}
	if err := conn.Model(migrations).
		Select(); err != nil {
		return nil, err
	}
	for i, item := range *migrations {
		fromPlace := &Place{}
		toPlace := &Place{}
		itm := &Item{}
		itmType := &ItemType{}
		conn.Model(itm).Where("uuid = ?0", item.ItemUUID).Select()
		conn.Model(itmType).Where("uuid = ?0", itm.TypeUUID).Select()
		conn.Model(fromPlace).Where("uuid = ?0", item.FromUUID).Select()
		conn.Model(toPlace).Where("uuid = ?0", item.ToUUID).Select()
		(*migrations)[i].Item = itm
		(*migrations)[i].Item.Type = itmType
		(*migrations)[i].From = fromPlace
		(*migrations)[i].To = toPlace
	}
	return migrations, nil
}

func (itm *Migration) GetAllMigrationsFromDevice(conn *pg.DB) (*[]Migration, error) {
	migrations := &[]Migration{}
	if err := conn.Model(migrations).
		Where("device_uuid.uuid = ?0", itm.UUID).
		Select(); err != nil {
		return nil, err
	}
	return migrations, nil
}
